src/bin/pgcopydb/ld_transform.c

```diff
@@ -1213,16 +1268,28 @@ streamLogicalTransactionAppendStatement(LogicalTransaction *txn,
 	}
 	else
 	{
-		if (txn->last != NULL)
+		if (coalesceStatements(txn->last, stmt))
 		{
-			/* update the current last entry of the linked-list */
-			txn->last->next = stmt;
+			/*
+			 * Deallocate new statement as it is merged into the
+			 * existing statment.
+			 */
+			free(stmt);
 		}
+		else
+		{
+			/* couldn't coalesce, append to the txn list as a new statement */
+			if (txn->last != NULL)
+			{
+				/* update the current last entry of the linked-list */
+				txn->last->next = stmt;
+			}
 
-		/* the new statement now becomes the last entry of the linked-list */
-		stmt->prev = txn->last;
-		stmt->next = NULL;
-		txn->last = stmt;
+			/* the new statement now becomes the last entry of the linked-list */
+			stmt->prev = txn->last;
+			stmt->next = NULL;
+			txn->last = stmt;
+		}
```

Comment:
I wonder if we should have `coalesceStatements` return a boolean that indicates if it could coalesce, have another function that only check if coalescing makes sense, or have the pointer management done within `coalesceStatements`.

At the moment I think that when adding the code to check that both statement (previous/next) share the same column target list we will prefer having a separate function to check that.

Reply:
I'm considering another approach: Call coalesce only for the DML case here in https://github.com/dimitri/pgcopydb/blob/a15881860dc18ad268c7bd8cae11dbe0055538dd/src/bin/pgcopydb/ld_transform.c#L1162

something like below,
```c
			if (canCoalesceLogicalTrasactionStatement(txn, stmt))
			{
				if (!coalesceLogicalTrasactionStatement(txn, stmt))
				{
					/* errors have already been logged */
					return false;
				}
			}
			else
			{
				(void) streamLogicalTransactionAppendStatement(txn, stmt);
			}
```